### PR TITLE
Add wheel slip plugin to vehicles

### PR DESCRIPTION
**Pull request by azeey**
The Wheel slip plugin dynamically adjusts the slip parameter of wheels in proportion to the speed of the vehicle. This is a better (i.e. more realistic) model of tires than the raw slip parameter in `<surface>` elements. More importantly, the plugin prevents the vehicle from sliding laterally when it is at rest. 

I have set the `<slip_compliance_lateral>` lateral parameter experimentally to maintain the current turning behavior of the vehicles. `<wheel_normal_force>` is the nominal normal force at the wheel. For most vehicles, this is calculated by dividing the weight of the vehicles equally among the wheels. For vehicles that have non-negligible weights in either the front or rear of the vehicle, those weights are assumed to be carried by only the front or rear wheels respectively.

## Requires:
- https://github.com/azeey/dart/pull/9
- https://github.com/ignitionrobotics/ign-physics/pull/56
- https://github.com/ignitionrobotics/ign-gazebo/pull/134

### TODO:
  - [x] CSIRO_DATA61_OZBOT_ATR_SENSOR_CONFIG_1
  - [x] CSIRO_DATA61_OZBOT_ATR_SENSOR_CONFIG_2
  - [x] CTU_CRAS_NORLAB_ABSOLEM_SENSOR_CONFIG_1
  - [x] CTU_CRAS_NORLAB_ABSOLEM_SENSOR_CONFIG_2

### CODE REVIEW

**Comment by nkoenig** [comment]
Resolved conflicts and added plugin to other launch files here: https://github.com/azeey/subt/pull/1

**Comment by JShep1** [comment]
Also tested all of the models out, I think @adlarkin 's comments sum it up. I tested out the `robotika_kloubak` the best I could and it seemed to be good.  I believe the `marble_hd2` oscillates when traveling forward due to the way that the tracks are modeled?  Perhaps this oscillation is exacerbated due to all of the various forces on each "wheel" interacting with each other?  Just a thought.  

Other than that, all seems to be good.  There's still slight drift/sliding when a vehicle is traveling up a slope or along a slope, it's a lot less drastic than before, and personally, I think it adds an element of realism (dust and tire material as well as other factors can introduce this sort of slipping in real life).  All vehicles come to a rest on a slope and don't drift.

**Comment by adlarkin** [comment]
> There's still slight drift/sliding when a vehicle is traveling up a slope or along a slope, it's a lot less drastic than before, and personally, I think it adds an element of realism (dust, rubble, and tire material as well as other factors can introduce this sort of slipping in real life).

I noticed this too, but I believe that slipping while a wheel is rotating is actually the correct behavior - @azeey since you probably have better physics knowledge than me, can you confirm that the slipping when moving behavior is correct/intended?

**Comment by azeey** [comment]
> marble_hd2 models oscillate back and forth significantly when it shouldn't (this is especially noticeable over flat terrain). I don't think that's related to the wheel slip plugin, though. It seems to be an issue I noticed with the model itself (the third point: #516 (review)). Since the wheel slip plugin modifies navigation behavior, the oscillation problem for marble_hd2 is now significantly worse/more noticeable.

Setting the radius of the rear wheels equal to the others seems to help with this ([my comment on #516](https://github.com/osrf/subt/pull/516#discussion_r475060169)). However, I did some additional tests on a slope and this vehicle doesn't seem to do well with the current Wheel Slip plugin parameters. I'll try to look into it.

> I noticed this too, but I believe that slipping while a wheel is rotating is actually the correct behavior - @azeey since you probably have better physics knowledge than me, can you confirm that the slipping when moving behavior is correct/intended?

Yes, this is the intended behavior. The amount of slipping will depend on the angular velocity of the wheel. When it's zero, there is no slip. This models the behavior of rubber tires more accurately than setting constant friction and slip parameters.

**Comment by azeey** [comment]
I have added the WheelSlip plugin to the remaining models and updated all the models to improve steering behavior, especially steering in place. To see the updated steering behavior, launch with `localModel:=true`. I have also updated the raw slip parameters in `<surface>` to account for the change in interpretation of the values made by azeey/dart#9. This ensures that the models behave as they used to even without the WheelSlip plugin. The expression I used for converting the older slip values to the new ones is `new_slip = old_slip/total_mass`. This expression is not an exact conversion between the old and the new slip parameters, but it's adequate for our purposes.

**Comment by azeey** [comment]
As @adlarkin encountered, the following robots won't work with `localModel:=true` because their `model.sdf` file is not in this repo. 

* All built-in X1 and X2 models
* CSIRO_DATA61_OZBOT_ATR_SENSOR_CONFIG_2
* CTU_CRAS_NORLAB_ABSOLEM_SENSOR_CONFIG_2
* EXPLORER_R2_SENSOR_CONFIG_2
* MARBLE_HD2_SENSOR_CONFIG_2
* MARBLE_HUSKY_SENSOR_CONFIG_2
* ROBOTIKA_FREYJA_SENSOR_CONFIG_2
* ROBOTIKA_KLOUBAK_SENSOR_CONFIG_2



Testing with their corresponding `CONFIG_1` robot should suffice until the models are updated on Fuel. For the built-in X1 and X2 models, test with X1_SENSOR_CONFIG_6 and X2_SENSOR_CONFIG_7

**Review by adlarkin — APPROVED** [review]
Outside of the models we can't test right now as mentioned by @azeey, I think all of the other models look good. Rotating in place seems to be pretty tight/constrained with these new parameters.

File: submitted_models/robotika_kloubak_sensor_config_1/launch/spawner.rb [hunk 1]
```diff
@@ -75,25 +75,25 @@ def spawner(_name, _modelURI, _worldName, _x, _y, _z, _roll, _pitch, _yaw)
         <plugin filename="libignition-gazebo-wheel-slip-system.so"
```
[the remark this answers is not in the document]
Comment (azeey) [reply submitted_models/robotika_kloubak_sensor_config_1/launch/spawner.rb:75]: Assuming your teleop is publishing to `Xa/cmd_vel`, you can use topic_tools/relay in ROS to send the same command to the front and rear wheels
```
rosrun topic_tools relay /Xa/cmd_vel /Xa/cmd_vel/front
rosrun topic_tools relay /Xa/cmd_vel /Xa/cmd_vel/rear
```

That seems to work well as long as the two axles are close to parallel.

[the remark this answers is not in the document]
Comment (adlarkin) [reply submitted_models/robotika_kloubak_sensor_config_1/launch/spawner.rb:75]: Thanks. Based on the testing I did with `topic_tools/relay`, it seems like the `robotika_kloubak` is good as well.

File: submitted_models/x1_sensor_config_6/launch/spawner.rb [hunk 2]
```diff
@@ -62,25 +62,25 @@ def spawner(_name, _modelURI, _worldName, _x, _y, _z, _roll, _pitch, _yaw)
       <plugin filename="libignition-gazebo-wheel-slip-system.so"
```
[the remark this answers is not in the document]
Comment (azeey) [reply submitted_models/x1_sensor_config_6/launch/spawner.rb:62]: I found a small fix that will make this better.

[the remark this answers is not in the document]
Comment (azeey) [reply submitted_models/x1_sensor_config_6/launch/spawner.rb:62]: https://github.com/osrf/subt/pull/569/commits/00e0cdb2ed06b97301622471bc4c30be6d5df110

**Comment by adlarkin** [comment]
Everything looks good to me after the change in [00e0cdb](https://github.com/osrf/subt/commit/00e0cdb2ed06b97301622471bc4c30be6d5df110).

**Comment by nkoenig** [comment]
All upstream releases have been made. I'll merge this sot that we can get a release out.

**Pull request merged**